### PR TITLE
Motor direction dialog for Dshot escs

### PR DESCRIPTION
**Pull request by limonspb**
A dialog to fix motor direction without necessity of going to blheli_s/32/jesc configurator.
Works through Dshot commands so appears only if Dshot protocol selected.

Firmware changes pull request:
https://github.com/betaflight/betaflight/pull/10372

Related Firmware bug for blocking dshot commands:
https://github.com/betaflight/betaflight/issues/10375
and it's fix by @mikeller:
https://github.com/betaflight/betaflight/pull/10406

image

### CODE REVIEW

File: locales/en/messages.json [hunk 1]
```diff
@@ -2579,6 +2579,43 @@
...
+        "message": "<strong>I understand the risks</strong>,<br />the propellers are removed."
+    },
+    "escDshotDirectionDialogExplanations": {
+        "message": "<strong>Information notice</strong><br />A motors will spin up once you select it and push \"Test\". The battery should be plugged in, correct ESC protocol should be selected."
+    },
+    "dialogEscDshotDirectionClose": {
+        "message": "Close"
+    },
+    "dialogEscDshotDirectionOpen": {
+        "message": "Fix motor direction"
```
Comment (Asizon) [inline locales/en/messages.json:2601]: Good job Lemon lets try it, I think it should be named "Motor direction" like dialog tittle.

**Comment by Asizon** [comment]
Tested @limonspb ! It works so good, only two things:
-Maybe autospin of 1200 is to high if we need to see the motor direction, can we made something to use 1100 or similar?
-An strange issue when click "Spin motor",  If you press it and move the mouse out of the dialog, you can release the mouse click and the motor keeps turning.

**Comment by Asizon** [comment]
I forgot another point:
-Spin button hover seems ok, but normal and reverse buttons have a missing hover style action

**Comment by limonspb** [comment]
> I forgot another point:
> -Spin button style seems ok when you click it, but normal and reverse buttons have a missing click style action

this is maybe something to think about...
None of the buttons in Configurator has style for when you click them.
I decided to add this "clicked" style for "spin motor" because it's a button user will push and keep pushing as long as he needs to spin a motor. Where as "direction normal/reversed" buttons are just normal buttons like every other button in the configurator.

**Comment by Asizon** [comment]
ok @limonspb now I sew what happens with this, i can see that the other yeellow button around configurator have a hover style,This is why I have not seen the normal and reverse buttons intuitive.

**Comment by limonspb** [comment]
> Tested @limonspb ! It works so good, only two things:
> -Maybe autospin of 1200 is to high if we need to see the motor direction, can we made something to use 1100 or similar?
> -An strange issue when click "Spin motor", If you press it and move the mouse out of the dialog, you can release the mouse click and the motor keeps turning.

@Asizon 
1. I think you are right about 1100. Because this is Dshot and 1100 should give some motor rotation speed anyway, shouldn't it?
2. I can't reproduce this on my windows. Could you please confirm? When i press and move the motor is still spinning, but it stops as soon as I release. Here is the code for that:
`        this._spinButton.on("mouseup **mouseleave**", () => {
            this._spinButton.removeClass("selected");
            this._spinMotor(-1);
        });`
as you can see there is "mouseleave"... Any ideas? Thanks!

**Comment by Asizon** [comment]
> > Probado @limonspb ! Funciona muy bien, solo dos cosas: 
> > -Tal vez el giro automático de 1200 sea demasiado alto si necesitamos ver la dirección del motor, ¿podemos hacer algo para usar 1100 o similar? 
> > -Un problema extraño al hacer clic en "Girar motor". Si lo presiona y mueve el mouse fuera del cuadro de diálogo, puede soltar el clic del mouse y el motor sigue girando.
> 
> @Asizon
> 
> 1. Creo que tienes razón sobre 1100. Porque esto es Dshot y 1100 debería dar alguna velocidad de rotación del motor de todos modos, ¿no?
> 2. No puedo reproducir esto en mis ventanas. ¿Podrías confirmar? Cuando presiono y muevo, el motor sigue girando, pero se detiene tan pronto como lo suelto. Aquí está el código para eso: 
>    `        this._spinButton.on("mouseup **mouseleave**", () => { this._spinButton.removeClass("selected"); this._spinMotor(-1); });`
>    como puede ver, hay "mouseleave" ... ¿Alguna idea? ¡Gracias!

Yes @limonspb it only happens if you release gently and don't move the pointer anymore

**Comment by Asizon** [comment]
Hover background color for bottons is #ffcc3e @limonspb

**Comment by limonspb** [comment]
> Hover background color for bottons is #ffcc3e @limonspb , maybe we can add it to all motor tab buttons?

I guess also for Blackbox tab? I think all the rest have a hover effect for buttons.

**Comment by Asizon** [comment]
Wow!Now looks really awesome, very intuitive navigating between buttons.
Spin button works as expected, i think now is a segure and perfect button.
Spining speed looks better now.
Thankyou for this great feature @limonspb

**Comment by mikeller** [comment]
Interesting. I made a proposal on how to change the protocol in https://github.com/betaflight/betaflight/pull/10372.

File: src/components/EscDshotDirection/EscDshotDirectionComponent.js [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        buffer = [];
+        buffer.push8(this._selectedMotor);
+        buffer.push8(DshotCommand.dshotCommands_e.DSHOT_CMD_SAVE_SETTINGS);
+        buffer.push8(DshotCommand.dshotCommandType_e.DSHOT_CMD_TYPE_INLINE);
+        MSP.send_message(MSPCodes.MSP2_SEND_DSHOT_COMMAND, buffer);
+    }
+
+    _setupSpinButtonEvents()
+    {
+        this._spinButton.on("mousedown", () => {
```
Comment (Asizon) [inline src/components/EscDshotDirection/EscDshotDirectionComponent.js:90]: @limonspb We need to add here these atributes for Android compatibility(else it doesnt works) `touchstart` and `touchend` like this:
```
    _setupSpinButtonEvents()
    {
        this._spinButton.on("mousedown touchstart", () => {
            this._spinButton.addClass("selected");
            this._spinMotor(this._selectedMotor);
        });

        this._spinButton.on("mouseup mouseout touchend", () => {
            this._spinButton.removeClass("selected");
            this._spinMotor(-1);
        });
    }
```

**Review by Asizon — CHANGES_REQUESTED** [review]
After some test on Android seems that motor spining button doesnt work as expected, with this solution should work.

**Comment by limonspb** [comment]
Firmware part is merged into master.
Temporarily sending DSHOT_CMD_TYPE_INLINE commands, not DSHOT_CMD_TYPE_BLOCKING
until blocking commands are fixed in bidirectional dshot.

**Comment by McGiverGim** [comment]
One suggestion if someone has time to do it: now that we have the "motors" tab with more features, I think we must in a future move all the motors configuration here:
- Move the servos test to the servos tab, to make some room here.
- Move the selection of number of motors, protocol, etc. from others tabs to the motors tab.
In this way I think is most intuitive. All the motor related stuff in the motors tab. Is only a suggestion if someone has time to fight against it. If someone is interested we can discuss it creating a new feature request to let this PR clean.

File: src/css/tabs/motors.css [hunk 3]
```diff
@@ -39,6 +39,31 @@
...
+    margin-right: 0px;
+    margin-bottom: 0px;
+    position: absolute;
+    right: 0px;
+    bottom: 0px;
+}
+
+.tab-motors #dialogEscDshotDirection {
+    width: 400px;
+    height:440px
```
Comment (haslinghuis) [inline src/css/tabs/motors.css:52]: Found one spacing issue ;-) 
```suggestion
    height: 440px
```

**Review by haslinghuis — APPROVED** [review]

**Comment by limonspb** [comment]
The main code is ready. I'm just adding some warning messages for when the number of motors in the mixer is 0 or when a non-dshot protocol is selected. So that user knew why he can't use this tool until he fixes the configuration.

File: src/js/tabs/motors.js [hunk 4]
```diff
@@ -403,16 +395,16 @@ TABS.motors.initialize = function (callback) {
...
                 drawGraph(gyro_helpers, gyro_data, samples_gyro_i);
-                for (var i = 0; i < 3; i++) {
+                for (let i = 0; i < 3; i++) {
                     if (Math.abs(gyro[i]) > Math.abs(gyro_max_read[i])) gyro_max_read[i] = gyro[i];
                 }
                 computeAndUpdate(gyro, gyro_data, gyro_max_read);
             }
 
             function computeAndUpdate(sensor_data, data, max_read) {
                 var sum = 0.0;
```
Comment (haslinghuis) [inline src/js/tabs/motors.js:405]:
```suggestion
                let sum = 0.0;
```

File: src/components/EscDshotDirection/Styles.css [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+.escDshotDirectionDialogInformationNotice {
+    font-size: 1.0em;
+    margin-bottom: 1.5em;
+    margin-top: 1.5em;
+}
+
+#escDshotDirectionDialog-MixerPreview {
+    width: 100%;
+    padding-top: 8px;
+	padding-bottom: 9px;
```
Comment (haslinghuis) [inline src/components/EscDshotDirection/Styles.css:52]: Padding is a little off in this file ;-)  Please use right amount of spaces for identation.

File: src/components/EscDshotDirection/Styles.css [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+.escDshotDirectionDialog-InformationNotice {
+    margin-top: 18px;
+}
+
+.escDshotDirectionDialog-StartButton {
+    width: 200px;
+    text-align: center;
+    margin: 8px;
+}
```
Comment (haslinghuis) [inline src/components/EscDshotDirection/Styles.css:158]: Please add newline here.

**Review by haslinghuis — CHANGES_REQUESTED** [review]
Great work @limonspb !!! Good quality. Some really minor comments.

**Review by haslinghuis — DISMISSED** [review]
LGTM!

**Comment by haslinghuis** [comment]
Accidently approved, status is WIP.

**Comment by limonspb** [comment]
I think this is ready for review and maybe merged after review is done?

This bug still exists:
https://github.com/betaflight/betaflight/issues/10375
so I had to use dshot inline commands 20 times in a loop to make sure they go through.

I think when the bug is fixed it's easier to replace inline with blocking commands than worry about merge conflicts in the future.

**Review by haslinghuis — DISMISSED** [review]
LGTM!

File: src/js/utils/DshotCommand.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    DSHOT_CMD_LED1_ON: 23, // BLHeli32 only
+    DSHOT_CMD_LED2_ON: 24, // BLHeli32 only
+    DSHOT_CMD_LED3_ON: 25, // BLHeli32 only
+    DSHOT_CMD_LED0_OFF: 26, // BLHeli32 only
+    DSHOT_CMD_LED1_OFF: 27, // BLHeli32 only
+    DSHOT_CMD_LED2_OFF: 28, // BLHeli32 only
+    DSHOT_CMD_LED3_OFF: 29, // BLHeli32 only
+    DSHOT_CMD_AUDIO_STREAM_MODE_ON_OFF: 30, // KISS audio Stream mode on/Off
+    DSHOT_CMD_SILENT_MODE_ON_OFF: 31, // KISS silent Mode on/Off
+    DSHOT_CMD_SIGNAL_LINE_TELEMETRY_DISABLE: 32,
```
Comment (mikeller) [inline src/js/utils/DshotCommand.js:34]: These two can be removed, they are being removed from the firmware as well: https://github.com/betaflight/betaflight/pull/10402

**Review by mikeller — COMMENTED** [review]
Looks good, some cosmetic remarks.

I think instead of updating the local copy of the openlayers library we should be ditching it and moving to the npm version.

One thing to keep in mind is that not all ESCs supporting Dshot support configuration through Dshot commands - see [Open ESC](https://github.com/AlkaMotors/AM32-MultiRotor-ESC-firmware) for an example.

Re having the wizard / individual methods, it would be nice to collect stats on their usage - but I can add this later in a separate pull request

**Comment by limonspb** [comment]
@mikeller, interesting note about not all ESCs supporting dshot configuration. Shoud I add it into "information notes"?
Like: Note that not all ESC firmwares support configuration through dshot commands. Please check with your ESC firmware.

**Comment by mikeller** [comment]
@limonspb: Yes, I think a notice like what you propose is appropriate - to tell users that this may not work with their hardware due to factors outside of our control.

**Comment by limonspb** [comment]
Should be tested with firmware from here until it will be merged into the firmware:
https://github.com/betaflight/betaflight/pull/10406

Sorry, guys, I hope it's a last "big" change in this PR. After switching to blocking Dshot commands I figured they can't be sent too often, right after motors got stopped. So had to implement an "MSP command queue" and some delays in the dialog.

**Review by haslinghuis — COMMENTED** [review]
I like the MSPCommandQueue idea. Maybe can use it for other things as well.

- What is pushPause() exactly doing? It does send a msp command continuous within a certain interval? If so better rename to pushContinuous() - the interval is evident by the argument.
- PurgeAndStop() is calling stop() but not clear();
- stopAllNow() is calling clear() but not purgeAndStop()

**Comment by limonspb** [comment]
> I like the MSPCommandQueue idea. Maybe can use it for other things as well.
> 
> * What is pushPause() exactly doing? It does send a msp command continuous within a certain interval? If so better rename to pushContinuous() - the interval is evident by the argument.
> * PurgeAndStop() is calling stop() but not clear();
> * stopAllNow() is calling clear() but not purgeAndStop()

Yeah the naming for these functions can be alittle messed up. Maybe should change them.

pushPause(milliseconds) is inserting a pause into the que. So that if you push more commands after that, they will be executed only after the pause.

PurgeAndStop() keeps the queue going and executing commands, but will kill the "interval timer" when there are no more commands. So maybe it should be "StopWhenEmpty()"?

stopAllNow() is a method in MotorDriver. Maybe it should be named as StopAllMotorsNow()
It clears the queue in case there are any commands left.
and then sends a single command to stop motors. So basically it stops motors almost immediately no matter what.

**Comment by haslinghuis** [comment]
@limonspb Thanks. Yes renaming will make it more clear to understand the code and abstracts it away (to be adaptable for more than only motor commands).
`pushPause()` --> `pushWithDelay()` or `pushWithInterval()` or `pushAfter()`;
 `StopAllMotors()` would be a better name.
`PurgeAndStop` --> `StopWhenEmpty()` could be called `flushAndCommit()` and should call `clear()` in the end

**Comment by limonspb** [comment]
Renamed MSPCommandQueue methods according to @haslinghuis  suggestions.
Added Betaflight firmware version check which I previously forgot.

**Review by haslinghuis — APPROVED** [review]
LGTM!

**Comment by AlkaMotors** [comment]
This will work with open ESC firmware (am32) now too as of version 1.64 so no need of a warning!

**Review by mikeller — APPROVED** [review]

**Pull request merged**